core/src/main/java/org/apache/accumulo/core/conf/Property.java

```diff
@@ -964,6 +964,11 @@ public enum Property {
       "1.3.5"),
   TABLE_ARBITRARY_PROP_PREFIX("table.custom.", null, PropertyType.PREFIX,
       "Prefix to be used for user defined arbitrary properties.", "1.7.0"),
+  TABLE_MAJC_OUTPUT_DROP_CACHE("table.compaction.major.output.drop.cache", "false",
+      PropertyType.BOOLEAN,
+      "Setting this property to true will call"
+          + "FSDataOutputStream.setDropBehind(true) on the major compaction output stream.",
+      "2.1.1"),
```

Comment:
Do we really need this? Why not just always set it, and let it be cached when read?

Also, this is a user-facing configuration addition (not strictly public API, but analogous in terms of forwards/backwards-compatibility issues and user expectations). We try to avoid those kinds of additions in patch releases. Can we justify adding it in 2.1.1?

Reply:
If the key extent aware configurer is a compelling feature, then it would require a new table prop for the use case @dlmarion outlined.  So I think  the decision on adding this prop hinges around whether the configurer change is made. From that perspective would it make sense to do both in the same PR?

Reply:
> Can we justify adding it in 2.1.1?

I'm not sure we need to. There are plenty of examples of properties being added in a patch release, including 4 of them added in version 1.9.3.

> Why not just always set it, and let it be cached when read?
> Making it configurable give a an escape hatch for that.

This is exactly my concern.  I think it may be safe to always do this, but I'm not 100% positive. The way that Hadoop implemented this, you can tell the OS via the FADVISE_DONTNEED flag that the OS can purge the file from the page cache when the read or write is done (in this case write). I read that this is done by setting that advice on the file descriptor and that each process (the DN in this case) has a file descriptor table. I don't know if re-opening the same file shortly after would re-use the same file descriptor number and I don't know if that advice may prevent caching of the file on a subsequent read operation. My guess is that it's safe to call it on the majc output file, but I don't know for sure.

Reply:
@dlmarion  the use case you mentioned sounds pretty compelling to me.  If the change to support it were made it would be nice to implement the use case on top of the new changes to see if actually works as intended.  That could help with decision on whether merging is worthwhile.

Reply:
> If the key extent aware configurer is a compelling feature, then it would require a new table prop for the use case @dlmarion outlined.

Would it require a new property if we didn't do anything with it in the default compaction configurer? For the use case that I described, I was thinking that a user would have to write their own [configurer](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/conf/Property.java#L1007) and [override](https://github.com/apache/accumulo/blob/f3d9903f8072efa50e75970d971595a66dc8affe/core/src/main/java/org/apache/accumulo/core/client/admin/compaction/CompactionConfigurer.java#L73) the table property based on the KeyExtent.

Reply:
> actually works as intended.

I have no issue  with the testing you suggested. However, I don't know whether we will be able to tell if the file was *actually* purged from the OS page cache as the FADVISE directive is advice.

Reply:
> For the use case that I described, I was thinking that a user would have to write their own [configurer](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/conf/Property.java#L1007)

Yeah that is what I was thinking, the change only support someone writing their own configurer for the use case you described.

Reply:
> I'm not sure we need to. There are plenty of examples of properties being added in a patch release, including 4 of them added in version 1.9.3.

We should think about it and try to provide a strong justification for it. Just because we've done it before doesn't mean it's a good idea in every case (FWIW, I pushed back on at least 2 of those in 1.9.3). I'm not saying we can't do it. I'm just suggesting that we give it due consideration first. The main things to consider is risk of destabilizing changes in a patch release due to newer features, and forward-compatibility expectations when rolling back (for example, what is a reasonable user expectation when rolling back and the property configured for 2.1.1 isn't recognized under 2.1.0).

In this case, a reasonable strong justification might be measurable performance impact. If the performance impact is negligible or merely hypothetical, or could be worse as easily as it could be better, then those are reasons to hold off. But, if it's a clear benefit, and is safe to rollback to 2.1.0, minimal low-risk code changes, then that could add up to be a good justification to include it in a patch release. For me, the question about the measurable impact remains an unanswered one, which leaves me with reservations for including it.

Reply:
An alternate approach would be to remove the property and make use of Property.TABLE_ARBITRARY_PROP_PREFIX ("table.custom."). We could create a defined property in FileCompactor (e.g. table.custom.compaction.major.output.drop.cache), which defaults to false and allow the custom CompactionConfigurer to set it to true.

Reply:
I opened #3088 to add the extent to the configurer.  While doing that I noticed it would also be nice to add it to the selector for symmetry.

Reply:
@dlmarion The arbitrary properties are intended to be a namespace for storage of properties for user-pluggable components, not for builtin framework configurations. For example, setting the volume chooser class is a builtin configuration, but some volume choosers may make use of arbitrary properties to control their behavior. I don't think that situation applies here. My understanding is that FileCompactor is a builtin component, part of the compaction framework, and not a user-pluggable component.

Reply:
That's fair, a better place for the defined Property is in [CompactionConfigurer](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/client/admin/compaction/CompactionConfigurer.java). That's user-pluggable via a table [property](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/conf/Property.java#L1007). I'm not sure why the CompactionConfigurer class isn't in the SPI package.

Reply:
In 74b0e2c I removed this property from Property.java to CompactionConfigurer. I'm going to mark this conversation as resolved as the original issue was the addition of the property.